---
layout: default
title: Past Research
---
		<div id="content">
			<div class="post">
				<h2 class="title">Past Research</h2>
    <h3>Human Microbiome: Inferring function from phylogeny</h3>
    <p class="meta">I was a post-doc funded by the Canadian Human Microbiome Intiative on "Modeling and mapping microbial diversity and function with marker genes, genomes and metagenomes" led by <a href="http://www.biochem.dal.ca/faculty/facultypages/doolittle/">Dr. Ford Doolittle</a> with direct supervision by co-PI <a href="http://kiwi.cs.dal.ca/~beiko/">Dr. Robert Beiko</a>. This reserch focused on better understanding the <a href="http://onlinelibrary.wiley.com/doi/10.1111/1574-6976.12035/abstract">microbiome as a community</a>, which led to the development of a novel and widely used new computational tool for inferring "virtual metagenomes" from 16S data, called <a href="http://www.nature.com/nbt/journal/v31/n9/abs/nbt.2676.html">PICRUSt</a> </p>  

    <h3>Environmental Metagenomics</h3>
    <p class="meta">During my first post-doc I was a member of the <a href="http://openwetware.org/wiki/ISEEM">iSEEM</a> research group supported by the <a href="http://www.moore.org/">Moore Foundation</a> led by <a href="http://bobcat.genomecenter.ucdavis.edu/mediawiki/index.php/Main_Page">Dr. Jonathan Eisen</a> for improving computational methods for metagenomics data. This included building databases of completed genomes (<a href="http://bioinformatics.oxfordjournals.org/content/28/14/1947.long">MicrobeDB</a>) and protein families (<a href="http://www.biomedcentral.com/1471-2105/13/264">SFams</a>), as well as applying a data reduction technique called Non-Negative Matrix Factorization for the first time on a metagenomic dataset from the ocean. </p>  

    <h3>Phylogenomics of Halophiles</h3>
    <p class="meta">Participated in the analysis of 8 haloarchaea genomes led my <a href="http://www.bme.ucdavis.edu/facciotti/">Marc Facciotti at UC Davis.</a> This research was published in PLoS ONE <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0041389">PLoS ONE</a>.</p>
    <h3>Open Science: BioTorrents</h3>
    <p class="meta">In general, I am interested in large datasets and improving the accessibility of results, software, and datasets in science. To improve the transfer speeds of data I developed <a href="http://www.biotorrents.net">BioTorrents</a>, a website that allows open sharing of scientific data using the popular peer-to-peer <a href="http://en.wikipedia.org/wiki/BitTorrent_%28protocol%29">BitTorrent</a> sharing program. A manuscript describing BioTorrents was <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0010071">published in PLoS One</a>. </p>  
 
				<h3>Genomic Islands</h3>
				<p class="meta">My PhD research was focused on identifying and characterizing <a href="http://en.wikipedia.org/wiki/Genomic_island">genomic islands (GIs)</a>. GIs are large regions (8-250 kb) that are thought to have been horizontally transferred in bacterial genomes. GIs are often linked to pathogenicity, antibiotic resistance, and adaptation to particular environments. During this research project I developed a novel GI prediction program, called IslandPick, that uses comparative genomics for GI identification. After <a href="http://www.ncbi.nlm.nih.gov/pubmed/18680607">evaluating the accurary of several other GI prediction programs</a>, I incorporated the most accurate methods into a single comprehensive website for GI prediction called <a href="http://www.pathogenomics.sfu.ca/islandviewer/">IslandViewer</a>. I also was involved in a collaborative project for the annotation and identification of GIs in a particaular <a href="http://www.ncbi.nlm.nih.gov/pubmed/19047519">virulent strain of Pseudomonas aeruginosa</a>. I also describe some interesting relationships between <a href="http://www.ncbi.nlm.nih.gov/pubmed/19956607">GIs, CRISPRs, and phage</a>. All of these are projects are outlined in my <a href="http://morganlangille.com/papers/Langille_2009_Thesis.pdf">PhD thesis</a>.</p>
				<h3>Standards development for flow cytometry data</h3>
				<p class="meta">My PhD program included research rotations before choosing a PhD supervisor for the remainder of my thesis research. One of these rotations was spent on developing standards for flow cytometry. I learned quite a bit about onlotology development (OWL) and file speficiations using XML and XSLT. More information is in the <a href="http://www.ncbi.nlm.nih.gov/pubmed/16901228">publication</a>.</p>
				<h3>Medical image segmentation algorithms</h3>
				<p class="meta">I dabbled in the field of medical image analysis, while taking an elective course with <a href="http://www.cs.sfu.ca/~hamarneh/">Dr. Hamarneh</a> during the start of my PhD. This research was focused on developing new segmentation algorithms (think magic wand tool in Photoshop) in 2D and 3D MRI images. This gave me a crash course in learning Matlab and C. The specifics on the research can be found in two publications: <a href="http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.69.5057">here</a> and <a href="http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.70.8575">here</a>. </p>
				<h3>Gene duplication in Drosophila</h3>
				<p class="meta">While completing my concurrent undergraduate Bachelor's of Science and Bachelor's of Computer Science degrees at UNB, I worked for two summers with <a href="http://www.unb.ca/fredericton/science/biology/Faculty/Clark.html">Dr. Clark</a> on identifying retrotransposed (DNA->RNA->DNA) gene duplications in <i>Drospophila melanogaster</i> and looking at gene expression differences between the parent and child copies using microarray data. More information about this project is explained in the <a href="http://www.ncbi.nlm.nih.gov/pubmed/17628393">publication</a>. </p>

			</div>
			
			
		<div style="clear: both;">&nbsp;</div>
		</div>
