<!DOCTYPE HTML>
<!--
	Arcana by HTML5 UP
	html5up.net | @n33co
	Free for personal and commercial use under the CCA 3.0 license (html5up.net/license)
-->
<html>
	<head>
		<title>Langille Lab @ Dalhousie University</title>
		<meta http-equiv="content-type" content="text/html; charset=utf-8" />
		<meta name="description" content="" />
		<meta name="keywords" content="" />
		<!--[if lte IE 8]><script src="css/ie/html5shiv.js"></script><![endif]-->
		<script src="js/jquery.min.js"></script>
		<script src="js/jquery.dropotron.min.js"></script>
		<script src="js/skel.min.js"></script>
		<script src="js/skel-layers.min.js"></script>
		<script src="js/init.js"></script>
		<noscript>
			<link rel="stylesheet" href="css/skel.css" />
			<link rel="stylesheet" href="css/style.css" />
			<link rel="stylesheet" href="css/style-wide.css" />
		</noscript>
		<!--[if lte IE 8]><link rel="stylesheet" href="css/ie/v8.css" /><![endif]-->
	</head>
	<body>

		<!-- Header -->
			<div id="header">
						
				<!-- Logo -->
					<h1><a href="index.html" id="logo">The Langille Lab <em>@ Dalhousie University</em></a></h1>
				
				<!-- Nav -->
					<nav id="nav">
						<ul>
							<li><a href="index.html">Home</a></li>
							<li><a href="members.html">Members</a></li>
							<li><a href="current-research.html">Current Research</a></li>
							<li class="current"><a href="past-research.html">Past Research</a></li>
							<li><a href="pubs.html">Publications</a></li>
							<li><a href="teaching.html">Teaching</a></li>
							<li><a href="contact.html">Contact Us</a></li>
						</ul>
					</nav>

			</div>
			
		<!-- Main -->
			<section class="wrapper style1">
				<div class="container">
					<div id="content">

						<!-- Content -->
					
							<article>
								<header>
									<h2>Past Research</h2>
								</header>
								
									<h3>Human Microbiome: Inferring function from phylogeny</h3>
									<p>I was a post-doc funded by the Canadian Human Microbiome Initiative on "Modelling and mapping microbial diversity and function with marker genes, genomes and metagenomes" led by <a href="http://www.biochem.dal.ca/faculty/facultypages/doolittle/" target="new">Dr. Ford Doolittle</a> with direct supervision by co-PI <a href="http://kiwi.cs.dal.ca/~beiko/" target="new">Dr. Robert Beiko</a>. This research focused on better understanding the <a href="http://onlinelibrary.wiley.com/doi/10.1111/1574-6976.12035/abstract" target="new">microbiome as a community</a>, which led to the development of a novel and widely used new computational tool for inferring "virtual metagenomes" from 16S data, called <a href="http://www.nature.com/nbt/journal/v31/n9/abs/nbt.2676.html" target="new">PICRUSt</a></p>
									<h3>Environmental Metagenomics</h3>
									<p>During my first post-doc I was a member of the <a href="http://openwetware.org/wiki/ISEEM" target="new">iSEEM</a> research group supported by the <a href="http://www.moore.org/" target="new">Moore Foundation</a> led by <a href="http://bobcat.genomecenter.ucdavis.edu/mediawiki/index.php/Main_Page" target="new">Dr. Jonathan Eisen</a> for improving computational methods for metagenomics data. This included building databases of completed genomes (<a href="http://bioinformatics.oxfordjournals.org/content/28/14/1947.long" target="new">MicrobeDB</a>) and protein families (<a href="http://www.biomedcentral.com/1471-2105/13/264" target="new">SFams</a>), as well as applying a data reduction technique called Non-Negative Matrix Factorization for the first time on a metagenomic dataset from the ocean.</p>
									<h3>Phylogenomics of Halophiles</h3>
									<p>Participated in the analysis of 8 haloarchaea genomes led my <a href="http://www.bme.ucdavis.edu/facciotti/" target="new">Marc Facciotti at UC Davis.</a> This research was published in PLoS ONE <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0041389" target="new">PLoS ONE</a>.</p>
									<h3>Open Science: BioTorrents</h3>
									<p>In general, I am interested in large datasets and improving the accessibility of results, software, and datasets in science. To improve the transfer speeds of data I developed <a href="http://www.biotorrents.net/" target="new">BioTorrents</a>, a website that allows open sharing of scientific data using the popular peer-to-peer <a href="http://en.wikipedia.org/wiki/BitTorrent_%28protocol%29" target="new">BitTorrent</a> sharing program. A manuscript describing BioTorrents was <a href="http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0010071" target="new">published in PLoS One</a>.</p>
									<h3>Genomic Islands</h3>
									<p>My PhD research was focused on identifying and characterizing <a href="http://en.wikipedia.org/wiki/Genomic_island" target="new">genomic islands (GIs)</a>. 
									GIs are large regions (8-250 kb) that are thought to have been horizontally transferred in bacterial 
									genomes. GIs are often linked to pathogenicity, antibiotic resistance, and adaptation to particular 
									environments. During this research project I developed a novel GI prediction program, called IslandPick, 
									that uses comparative genomics for GI identification. After <a href="http://www.ncbi.nlm.nih.gov/pubmed/18680607" target="new">evaluating 
									the accurary of several other GI prediction programs</a>, I incorporated the most accurate methods into 
									a single comprehensive website for GI prediction called <a href="http://www.pathogenomics.sfu.ca/islandviewer/" target="new">IslandViewer</a>. 
									I also was involved in a collaborative project for the annotation and identification of GIs in a 
									particular <a href="http://www.ncbi.nlm.nih.gov/pubmed/19047519" target="new">virulent strain of Pseudomonas aeruginosa</a>. 
									I also describe some interesting relationships between <a href="http://www.ncbi.nlm.nih.gov/pubmed/19956607" target="new">GIs, CRISPRs, and phage</a>. 
									All of these are projects are outlined in my <a href="http://morganlangille.com/papers/Langille_2009_Thesis.pdf" target="new">PhD thesis</a>.</p>
									<h3>Standards development for flow cytometry data</h3>
									<p>My PhD program included research rotations before choosing a PhD supervisor for the remainder of my thesis research. One of these rotations was spent on developing standards for flow cytometry. I learned quite a bit about ontology development (OWL) and file specifications using XML and XSLT. More information is in the <a href="http://www.ncbi.nlm.nih.gov/pubmed/16901228" target="new">publication</a>.</p>
									<h3>Medical image segmentation algorithms</h3>
									<p>I dabbled in the field of medical image analysis, while taking an elective course with <a href="http://www.cs.sfu.ca/~hamarneh/" target="new">Dr. Hamarneh</a> during the start of my PhD. This research was focused on developing new segmentation algorithms (think magic wand tool in Photoshop) in 2D and 3D MRI images. This gave me a crash course in learning Matlab and C. The specifics on the research can be found in two publications: <a href="http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.69.5057" target="new">here</a> and <a href="http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.70.8575" target="new">here</a>.</p>
									<h3>Gene duplication in Drosophila</h3>
									<p>While completing my concurrent undergraduate Bachelor's of Science and Bachelor's of Computer Science degrees at UNB, I worked for two summers with <a href="http://www.unb.ca/fredericton/science/biology/Faculty/Clark.html" target="new">Dr. Clark</a> on identifying retrotransposed (DNA-&gt;RNA-&gt;DNA) gene duplications in <i>Drospophila melanogaster</i> and looking at gene expression differences between the parent and child copies using microarray data. More information about this project is explained in the <a href="http://www.ncbi.nlm.nih.gov/pubmed/17628393" target="new">publication</a>.</p>
								
							</article>
				
					</div>
				</div>
			</section>			

		<!-- Footer -->
			<div id="footer">
				<div class="container">
					<div class="row">
						<section class="3u 6u(2) 12u$(3)">
							<h3>Research Related</h3>
							<ul class="links">
								<li><a href="http://dalmug.org" target="new">DalMUG</a></li>
								<li><a href="http://picrust.github.com/" target="new">PICRUSt</a></li>
								<li><a href="http://www.pathogenomics.sfu.ca/islandviewer/query.php" target="new">IslandViewer</a></li>
								<li><a href="http://www.biotorrents.net" target="new">BioTorrents</a></li>
								<li><a href="https://github.com/mlangill/microbedb/" target="new">MicrobeDB</a></li>
							</ul>
						</section>
						<section class="3u 6u$(2) 12u$(3)">
							<h3>More Links</h3>
							<ul class="links">
								<li><a href="http://cgeb-imr.ca" target="new">CGEB-IMR</a></li>
								<li><a href="http://cgeb.dal.ca" target="new">CGEB</a></li>
								<li><a href="http://pharmacology.medicine.dal.ca" target="new">Dept. of Pharmacology</a></li>
								<li><a href="http://medicine.dal.ca" target="new">Faculty of Medicine</a></li>
								<li><a href="http://dal.ca" target="new">Dalhousie University</a></li>
							</ul>
						</section>
						<section class="6u 12u(2)">
							<a class="twitter-timeline" href="/BetaScience/timelines/552166510979194880" data-widget-id="552167351207358466" data-tweet-limit="5">Langille Lab</a>
							<script>!function(d,s,id){var js,fjs=d.getElementsByTagName(s)[0],p=/^http:/.test(d.location)?'http':'https';if(!d.getElementById(id)){js=d.createElement(s);js.id=id;js.src=p+"://platform.twitter.com/widgets.js";fjs.parentNode.insertBefore(js,fjs);}}(document,"script","twitter-wjs");</script>
						</section>
					</div>
				</div>

				<!-- Icons -->
					<ul class="icons">
						<li><a href="http://twitter.com/betascience" class="icon fa-twitter" target="new"><span class="label">Twitter</span></a></li>
						<li><a href="http://www.facebook.com/morgan.langille" class="icon fa-facebook" target="new"><span class="label" target="new">Facebook</span></a></li>
						<li><a href="http://github.com/mlangill" class="icon fa-github" target="new"><span class="label" target="new">GitHub</span></a></li>
						<li><a href="http://www.linkedin.com/in/morganlangille" class="icon fa-linkedin" target="new"><span class="label" target="new">LinkedIn</span></a></li>
						<li><a href="http://www.mendeley.com/profiles/morgan-langille" class="icon fa-file-text" target="new"><span class="label" target="new">Mendeley</span></a></li>
					</ul>

				<!-- Copyright -->
					<div class="copyright">
						<ul class="menu">
							<li>&copy; Langille Lab. All rights reserved</li><li>Design: <a href="http://html5up.net">HTML5 UP</a></li>
						</ul>
					</div>

			</div>

	</body>
</html>